### DISCUSSION TITLE
How to custom the response body for JWT and rate-limiting plugin?

### DISCUSSION THREAD

**Discussion by XJShane** [Kong/kong#7938] [Help]
Hi i want to custom the response body when JWT or rate-limiting plugin return, to compatible with older programs, for example:
{
    "status": 401,
    "msg": "token expired"
}

{
    "status": 429,
    "msg": "rate limit exceed"
}

kong version: 2.2.0

I'm not sure that modifying the source code for JWT and rate-limiting plugin is a good idea, if there's no other way.

**Comment by tyree731** [comment]
These plugins run in Kong/openresty's `access` phase, meaning that after they call `kong.response.exit`, you should be able to customize the returned response by running a custom plugin with an appropriate `header_filter` and `body_filter` phase.

**Comment by XJShane** [comment]
Hi @tyree731 i also need to get the response data for JWT plugin returned, i have already write a custom plugin, but after the JWT plugin,  the `access` phase in custom plugin not executed, but the response data only can get in this phase. I think the reason is that JWT `access` phase has called `kong.response.exit`, when the priority less than JWT plugin, it dont't executed the  `access` phase anymore.

**Comment by tyree731** [comment]
When you say "the response data for JWT plugin", do you mean the above referenced response body:

    {
      "status": 401,
      "msg": "token expired"
    }

If so, that body will be available if your plugin operates in the `body_filter` or `response` plugin phases. Perhaps an example would help? We have a custom plugin that authorizes users based on whether they have access to a given service. If the user does not have access, we run code that looks something like like:

    local forbidden_response = function(response_message)
      return kong.response.exit(403, response_message, {["Content-Type"] = "text/plain"})
    end

Separately we have a plugin that, when being requested from our websites, templates plaintext responses from downstream services to make those responses match our company's design guidelines. That plugin has a `header_filter` phase which checks to see if we will be templating the response, a check which uses request information and the response "Content-Type" header. If we will be templating the response, the `header_filter` portion of the plugin clears the response "Content-Length" header, sets an appropriate "Content-Type" header on the response, and sets a request context variable which indicates that the `body_filter` phase should run. In the `body_filter` phase we buffer the entire response body and, once that buffering is complete, we template the response body and send it out.

This response template plugin works on the body set by the `kong.response.exit` call. You should be able to do something similar for your example.
